### ISSUE TITLE
Incompatible licenses

### ISSUE THREAD

**Issue by Palakis** [Palakis/obs-ndi#4]
The source code is currently licensed under GPL 2.0. The NDI SDK is under a proprietary license (NewTek's EULA) but is not included in the source code of obs-ndi.

However :
- the code of obs-ndi makes use of the headers provided in the NDI SDK
- the code of obs-ndi is linked at compile-time to the NDI library
- the compiled releases of obs-ndi are distributed with a copy of NewTek's NDI libraries

So in it's current state, the plugin is not really GPL and hence can't legally be used within OBS Studio.

**Comment by stephanel** [comment]
This comment from HellGore is full of good ideas : https://github.com/CasparCG/Server/issues/379#issuecomment-255735201
- Replicating the API signatures and using DLL calls to use the library solves the headers issue (the headers of NewTek's NDI SDK are considered to be proprietary code)
- Having NewTek distribute NDI redistributables would solve the binary distribution issue (if NewTek makes the decision to release redistributables)
- But applying the System Library exception of the GPL to NDI libraries isn't possible.

**Comment by stephanel** [comment]
I believed the best license for obs-ndi would be LGPL v2.1.
The LGPL is more permissive regarding the use of proprietary software, and is compatible with the GPL v2.0 used by OBS Studio.

**Comment by Palakis** [comment]
Switching to LGPL does not seem possible. Used in OBS Studio (which is under GPL), the plugin would be considered to be a combined work with OBS, and the plugin would have to be under GPL 2.1. 

But the GPL is very strict regarding the use of proprietary libraries...

**Comment by Palakis** [comment]
Regarding the use of proprietary libraries, it seems like the GPL even forbids run-time linking (using LoadLibrary/dlopen, with type and function definitions wrote under an OSS license, and without linking at compile-time with a proprietary lib)...

**Comment by Palakis** [comment]
I switched the code to LGPL 2.1. The license may change again in the future, but as of now the GPL 2.0 isn't compatible with how the plugin works with proprietary libraries.

**Comment by sesse** [comment]
Switching your plugin to LGPL 2.1 doesn't do anything; LGPL 2.1 is [i]compatible[/i] with GPL only in the typical GPL-sense that you can [i]treat it as GPL[/i]. So once obs-ndi gets loaded into OBS, it automatically becomes GPL, and thus cannot link to NDI.

FSF's position in the GPL FAQ is that you can make non-GPL plugins for free software if they run in a separate process. (https://www.gnu.org/licenses/gpl-faq.html#GPLAndPlugins)

Another alternative would be to make an explicit exception for NDI, but you'd need to get permission of every copyright holder of code in OBS and all GPL code it includes/links to, which would seem impractical.

**Comment by Palakis** [comment]
I totally and completely agree with you that the LGPL 2.1 doesn't do anything regarding the plugin's use with OBS.
However, I switched to this license to "solve" the plugin's mandatory integration with proprietary components (the NDI SDK in this case), even if it doesn't solve the compatibility with OBS Studio's license.
Regarding the possible alternative of adding a linking exception in OBS for the NDI library, one of OBS' project leaders (dodgepong) said on IRC that they want to avoid linking exceptions as much as possible.

I developed obs-ndi for my own needs and wanted to share it under an free software license with the hope that it will be useful to other people. But I did not expect so many license issues when I started working on it, and I don't have time to work on ways to improve the GPL/LGPL compatibility of obs-ndi.
Right now, binary builds are only available through this repo's Releases section, and the plugin hasn't been published on the OBS forum.

People interested in having NDI integrated in OBS can contribute to [dalegaard's work on NDI support through IPC](https://github.com/dalegaard/obs-studio/tree/feature_ndi).
Since the source code of my NDI plugin is free software, contributors can take bits of it and put them in dalegaard's work.
